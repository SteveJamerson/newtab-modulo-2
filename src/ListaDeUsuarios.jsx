import React, { useEffect, useState } from "react";
import "./listadeusuarios.css";
import axios from "axios";

//Pegando as informações da API pelo GET
const ListaDeUsuarios = () => {
  const [_infos, _setInfos] = useState([]);
  const [infos, setInfos] = useState([]);
  useEffect(() => {
    axios
      .get("https://www.mocky.io/v2/5d531c4f2e0000620081ddce", {
        method: "GET",
      })
      .then((resposta) => {
        setInfos(resposta.data);
        _setInfos(resposta.data);
      })
  }, []);

  // Mock com lista de cartõs para teste
  const cards = [
    // cartão válido
    {
      card_number: "1111111111111111",
      cvv: 789,
      expiry_date: "01/18",
    },
    // cartão inválido
    {
      card_number: "4111111111111234",
      cvv: 123,
      expiry_date: "01/20",
    },
  ];

  const filter = (event) => {
    const value = event.target.value;
    const _new = _infos.filter((i) => i.id.toString().includes(value));
    setInfos(_new);
  };

  // Função para pegar a escolha do cartão do input select
  const escolhaDoCartao = (event) => {
    setValorCartao(event.target.value);
  };

  // Ações dos modals
  const [abrirPagamento, setAbrirPagamento] = useState("none"); // Para abrir modal de pagamento
  const [pegarUsuario, setPegarUsuario] = useState(""); // Para pegar o nome do usuário
  const [abrirPagou, setAbrirPagou] = useState("none"); // Para abrir modal com recibo de pagamento
  const [abrirNaoRecebeu, setAbrirNaoRecebeu] = useState(""); // Para msg de erro de pagamento
  const [valorCartao, setValorCartao] = useState("1"); // Para pegar o cartão escolhido para pagamento
  const [valorDinheiro, setValorDinheiro] = useState(""); // Para pegar o valor de pagamento digitado
  const [validarCampo, setValidarCampo] = useState("none"); // Para validar campo de valor digitado

  // Função para abrir o modal de pagamento do usuário
  const modalPagar = (name) => {
    setAbrirPagamento("flex");
    setPegarUsuario(name);
  };

  // Função que abre o modal de recibo de pagamento
  const modalPagou = () => {
    if (valorDinheiro === "") {
      setValidarCampo("flex");
    } else {
      if (valorCartao === "1") {
        setAbrirNaoRecebeu("");
      } else {
        setAbrirNaoRecebeu("não");
      }
      setAbrirPagamento("none");
      setAbrirPagou("flex");
      setValorDinheiro("");
      setValidarCampo("none");
    }
  };

  // Função para fechar o modal do recibo de pagamento
  const fecharModal = () => {
    setAbrirPagou("none");
  };

  // Função para validar campo de valor para pagamento do usuário
  const valorInput = (event) => {
    const value = event.target.value.replace(/\D/g, "");
    const amount = new Intl.NumberFormat("pt-BR", {
      style: "currency",
      currency: "BRL",
    }).format(value / 100);
    setValorDinheiro(amount);
    setValidarCampo("none");
  };

  // Renderizando na tela as informações recebidas da API
  return (
    <section className="container">
      <label htmlFor="search">Buscar usuário</label>
      <input
        className="field"
        type="search"
        name=""
        id="search"
        onChange={filter}
        placeholder="Pesquise por: Nome, ID, Username"
      />
      <div className="row">
        {infos.map((item) => (
          <div className="card" key={item.index}>
            <div className="card-body">
              <img className="thumbnail" src={item.img} alt="Foto do usuário" />
              <div className="infos">
                <p className="title">{item.name}</p>
                <ul className="description">
                  <li>
                    <b>ID:</b> {item.id}
                  </li>
                  <li>
                    <b>Username:</b> {item.username}
                  </li>
                </ul>
              </div>
              <button
                className="modal-pay"
                onClick={() => {
                  modalPagar(item.name);
                }}
              >
                Pagar
              </button>
            </div>
          </div>
        ))}

        {/*--------------------------------Abrir Modal de pagamento----------------------------------*/}
        <div className="modal" style={{ display: abrirPagamento }}>
          <div className="modal-content">
            <p className="modal-header">
              Pagamento para <span>{pegarUsuario}</span>
            </p>
            <div className="modal-body">
              <input
                className="field"
                thousandSeparator={true}
                value={valorDinheiro}
                onChange={valorInput}
                placeholder="R$ 0,00"
              />
              <small className="field-error" style={{ display: validarCampo }}>
                Campo obrigatório
              </small>
              <select
                className="field"
                value={valorCartao}
                onChange={escolhaDoCartao}
              >
                <option value="1">
                  Cartão com final {cards[0].card_number.substr(-4)}
                </option>
                <option value="2">
                  Cartão com final {cards[1].card_number.substr(-4)}
                </option>
              </select>
            </div>
            <div className="modal-footer">
              <button
                onClick={() => {
                  modalPagou();
                }}
              >
                Pagar
              </button>
            </div>
          </div>
        </div>

        {/*------------------------------Abrir Modal de recibo de pagamento--------------------------------*/}
        <div className="modal" style={{ display: abrirPagou }}>
          <div className="modal-content">
            <div className="modal-header ">
              <p>Recibo de pagamento</p>
            </div>
            <div className="modal-body">
              <p>
                O Pagamento <b>{abrirNaoRecebeu}</b> foi concluído com sucesso
              </p>
            </div>
            <div className="modal-footer">
              <button
                onClick={() => {
                  fecharModal();
                }}
              >
                Fechar
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};

export default ListaDeUsuarios;
