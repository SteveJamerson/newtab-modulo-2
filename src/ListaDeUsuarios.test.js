import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import ListaDeUsuarios, {} from "./ListaDeUsuarios";

import axios from "axios";

const API = [
  {
    id: 1001,
    name: "Eduardo Santos",
    img: "https://randomuser.me/api/portraits/men/9.jpg",
    username: "@eduardo.santos",
  },
  {
    id: 1002,
    name: "Marina Coelho",
    img: "https://randomuser.me/api/portraits/women/37.jpg",
    username: "@marina.coelho",
  },
];

jest.mock("axios", () => ({
  get: jest.fn((..._) => {
    return new Promise((resolve) => {
      resolve(true);
    });
  }),
}));

describe("Lista de usuarios", () => {
  beforeEach(async () => {
    axios.get.mockResolvedValue({ data: API });
    render(<ListaDeUsuarios />);
    await screen.findByText('Buscar usuÃ¡rio');
  })

  it("should loading", () => {
    render(<ListaDeUsuarios />);
    const test = screen.getByTestId("loading");
    expect(test).toBeTruthy();
    expect(test).toMatchSnapshot();
  });

  it("should container", async () => { 
    const test = screen.getByTestId("container");
    expect(test).toBeTruthy();
    expect(test).toMatchSnapshot();
  });

  it("should filter", async () => {
    const filter = screen.getByTestId("filter")
    fireEvent.change(filter, { target: { value: '10' } })
  });

  it("should open and close modal payment", async () => {
    
    const open = screen.getByTestId("modal-pay-1")
    fireEvent.click(open)
    const close = screen.getByTestId("modal-pay-close")
    fireEvent.click(close)
  });
});
