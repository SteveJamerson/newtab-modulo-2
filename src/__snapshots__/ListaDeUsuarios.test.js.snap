// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lista de usuarios should container 1`] = `
<section
  class="container"
  data-testid="container"
>
  <label
    for="search"
  >
    Buscar usuário
  </label>
  <input
    class="field"
    data-testid="filter"
    id="search"
    name=""
    placeholder="Pesquise por: Nome, ID e Username."
    type="search"
  />
  <div
    class="row"
  >
    <div
      class="card"
      data-testid="card-0"
    >
      <div
        class="card-body"
      >
        <img
          alt="Foto do usuário"
          class="thumbnail"
          src="https://randomuser.me/api/portraits/men/9.jpg"
        />
        <div
          class="infos"
        >
          <p
            class="title"
            data-testid="title-0"
          >
            Eduardo Santos
          </p>
          <ul
            class="description"
          >
            <li>
              <b>
                ID:
              </b>
               
              1001
            </li>
            <li>
              <b>
                Username:
              </b>
               
              @eduardo.santos
            </li>
          </ul>
        </div>
        <button
          class="modal-pay"
          data-testid="modal-pay-0"
        >
          Pagar
        </button>
      </div>
    </div>
    <div
      class="card"
      data-testid="card-1"
    >
      <div
        class="card-body"
      >
        <img
          alt="Foto do usuário"
          class="thumbnail"
          src="https://randomuser.me/api/portraits/women/37.jpg"
        />
        <div
          class="infos"
        >
          <p
            class="title"
            data-testid="title-1"
          >
            Marina Coelho
          </p>
          <ul
            class="description"
          >
            <li>
              <b>
                ID:
              </b>
               
              1002
            </li>
            <li>
              <b>
                Username:
              </b>
               
              @marina.coelho
            </li>
          </ul>
        </div>
        <button
          class="modal-pay"
          data-testid="modal-pay-1"
        >
          Pagar
        </button>
      </div>
    </div>
    <div
      class="modal"
      style="display: none;"
    >
      <div
        class="modal-content"
      >
        <form>
          <p
            class="modal-header"
          >
            Pagamento para 
            <span />
            <button
              class="close"
              data-testid="modal-pay-close"
            >
              <i
                class="close-icon"
              />
              <i
                class="close-icon"
              />
            </button>
          </p>
          <div
            class="modal-body"
          >
            <input
              class="field"
              placeholder="R$ 0,00"
              value=""
            />
            <small
              class="field-error"
              style="display: none;"
            >
              Campo obrigatório
            </small>
            <select
              class="field"
            >
              <option
                value="1"
              >
                Cartão com final 1111
              </option>
              <option
                value="2"
              >
                Cartão com final 1234
              </option>
            </select>
          </div>
          <div
            class="modal-footer"
          >
            <button
              type="submit"
            >
              Pagar
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="modal"
      style="display: none;"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header "
        >
          <p>
            Recibo de pagamento
          </p>
        </div>
        <div
          class="modal-body"
        >
          <p>
            O Pagamento 
            <b />
             foi concluído com sucesso
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button>
            Fechar
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`Lista de usuarios should loading 1`] = `
<div
  class="loading"
  data-testid="loading"
/>
`;
